### PR TITLE
Add access to MPTrie data

### PR DESCRIPTION
**Pull request by gennadylaventman**
Added support for data proof to `ledger_request_handler.go` and `http.go` from REST side and to `db.go` and `ledger_query_procesor.go` from internal implementation side.

Signed-off-by: Gennady Laventman <gennady@il.ibm.com>

### CODE REVIEW

File: internal/httphandler/ledger_request_handler.go
Comment (tock-ibm) [inline internal/httphandler/ledger_request_handler.go:47]: This will break SDK, need to reflect it there as well

File: internal/httphandler/ledger_request_handler.go
Comment (tock-ibm) [reply internal/httphandler/ledger_request_handler.go:41]: Again, why blockID and not key version?

File: internal/httphandler/ledger_request_handler.go
Comment (tock-ibm) [inline internal/httphandler/ledger_request_handler.go:42]: Is it possible to use this with a default `deleted=false` ?
deleted seems to be rare enough so that we allow a default for the frequent case.

File: pkg/constants/http.go
Comment (tock-ibm) [inline pkg/constants/http.go:40]: I think that if the user make a request to `/ledger/proof/tx` `/ledger/proof/tx/` `/ledger/proof/data` or `/ledger/proof/data/` he will get an unformatted error response content from within `gorilla/mux` which is not interpreted very well by the SDK.

File: internal/bcdb/db.go
Comment (gennadylaventman) [reply internal/bcdb/db.go:64]: Interesting idea. Let me explain original idea - block number used as time marker. So, once user ask for proof for key K inside block N, it doesn't mean that this key was updated in block N, it just mean that at the time of block N creation, value associated with key K was value with hash located at the end of proof.

File: internal/httphandler/ledger_request_handler.go
Comment (gennadylaventman) [reply internal/httphandler/ledger_request_handler.go:47]: Sure, next step will update sdk. Lets tag it after that to 0.1.1

File: internal/httphandler/ledger_request_handler.go
Comment (gennadylaventman) [reply internal/httphandler/ledger_request_handler.go:42]: Have to thinks about it...

File: pkg/constants/http.go
Comment (gennadylaventman) [reply pkg/constants/http.go:40]: It will get default 404 response. I can add multiple invalid handlers to take care of following cases: 
1. ` /ledger/proof/tx`
2. `/ledger/proof/data`
3. `/ledger/proof/data/{blockId}`
4. `ledger/proof/data/{blockId}/{dbname}`
But once user uses sdk, no such thing is possible. wdut?

Comment (tock-ibm) [reply pkg/constants/http.go:40]: I think it makes sense to cover all the corners, after all, the REST spec is public and user/community can write her own SDK.... especially in other languages, like Java, Node, who knows...

File: internal/bcdb/db.go
Comment (cendhu) [reply internal/bcdb/db.go:64]: @gennadylaventman I think it should be the version as mentioned by @tock-ibm For example, multiple transactions in a block can blindly write to a KV pair. The end value of the KV pair is the value written by the last transactions. However, the provenance store would return all writes performed to the KV pair. In order to get a proof for any data, the user can pass the block number and the version. 

Now, I wonder whether we take care of all blind writes within a block. I think, we do not. Am I right?

File: internal/httphandler/ledger_request_handler.go
Comment (gennadylaventman) [reply internal/httphandler/ledger_request_handler.go:39]: Done

File: internal/httphandler/ledger_request_handler.go
Comment (gennadylaventman) [reply internal/httphandler/ledger_request_handler.go:41]: Done

File: internal/httphandler/ledger_request_handler.go
Comment (gennadylaventman) [reply internal/httphandler/ledger_request_handler.go:42]: Done

File: internal/httphandler/ledger_request_handler.go
Comment (gennadylaventman) [reply internal/httphandler/ledger_request_handler.go:47]: Because sdk call to same `constants.URLTxProof(blockNum, txIndex)` function, new tag after merge will be enough. In any case, opened new issue to update sdk: https://github.com/IBM-Blockchain/bcdb-sdk/issues/29

File: internal/bcdb/db.go
Comment (gennadylaventman) [reply internal/bcdb/db.go:64]: We are not, because it will produce multiple trie roots for same block, at least as number of tx in block. It is doable, but will require same change it integration code and will increase block header size. We had discussion with @tock-ibm  and for now, at least, we going with block number. I already have issue to store metadata hash with value, adding this, I think, will solve version problem.

Comment (cendhu) [reply internal/bcdb/db.go:64]: How about we allow only one write to a key per block? In general, blind writes are very rare. Even if blind writes are done by the user, if we allow only one blind write to a key per block, we can prove all states present in the provenance store. Having a root hash per transaction would complicate things more.

Comment (gennadylaventman) [reply internal/bcdb/db.go:64]: I don't think root hash per tx is related to amount of writes to same key in block because each write to any key updates root hash. If we handle tx as one atomic update, each tx with writes updates root hash. This is why I mentioned root hash per tx. We had discussion with Yoav and we decided to keep current approach, just to store metadata hash with value hash and to make following changes to REST API `ledger/proof/data/db/key?block={blockNum}&deleted={true|false}`

Comment (cendhu) [reply internal/bcdb/db.go:64]: Discussed offline. We will make changes in MVCC to allow only one write to a given key per block.

File: internal/bcdb/ledger_query_procesor.go
Comment (cendhu) [inline internal/bcdb/ledger_query_procesor.go:134]: `GetHeader()` instead of `Get()`.

File: internal/bcdb/ledger_query_procesor.go
Comment (cendhu) [inline internal/bcdb/ledger_query_procesor.go:125]: `isDeleted` instead of `deleted`

File: internal/bcdb/ledger_query_procesor_test.go
Comment (cendhu) [inline internal/bcdb/ledger_query_procesor_test.go:357]: As this is an exact duplicate of the same function in the `committer.go`, we can extract the common part of an exported function and then call it here and in the committer package

File: internal/bcdb/ledger_query_procesor_test.go
Comment (cendhu) [inline internal/bcdb/ledger_query_procesor_test.go:434]: export the same function in `committer.go` and call it here.

File: internal/httphandler/ledger_request_handler.go
Comment (cendhu) [inline internal/httphandler/ledger_request_handler.go:55]: Not related to this PR. We need to do this for other handlers too.

File: pkg/constants/http_test.go
Comment (cendhu) [inline pkg/constants/http_test.go:88]: I am okay with the usage of `block` rather than `block_number`. To maintain consistency, we need to use `block` for other HTTP requests too. I think any of us can address this in a separate PR.

File: protos/query.proto
Comment (cendhu) [inline protos/query.proto:100]: `is_deleted`

**Review by cendhu — COMMENTED** [review]
A few minor comments. Otherwise, LGTM.

File: pkg/constants/http_test.go
Comment (gennadylaventman) [reply pkg/constants/http_test.go:88]: Agreed. I will open issue and lets discuss it.

https://github.com/IBM-Blockchain/bcdb-server/issues/111

**Review by cendhu — APPROVED** [review]
Thanks, @gennadylaventman

**Pull request merged**
